### ISSUE TITLE
Consider PackageFolders that abstract layout for cache, workspaces

### ISSUE THREAD

**Issue by memsharded** [conan-io/conan#3559]
Now, the layout of a conan package is cache is directly obtained from ``ClientCache``, which is hardcoded into many processes and algorithms.

It might be useful to create an abstract ``PackageFolders`` or similar, that abstract that. It will have a concrete ``CachePackageFolders`` which will be an indirection over the ``ClientCache``, but it could be replaced by a ``WorkspacePackageFolders``, for workspaces.

It could also enable improvements in the local workflow, and allow other cache layouts, and other operations (e.g. ``conan download`` directly to user folder)

**Comment by jgsogo** [comment]
Current `ConanClient` class has this initializator ([link](https://github.com/conan-io/conan/blob/develop/conans/client/client_cache.py#L41)):

```python
class ClientCache(SimplePaths):
    def __init__(self, base_folder, store_folder, output):
```
isn't it enough? Maybe we only need a better naming.

**Comment by memsharded** [comment]
I am not talking about that. The ClientCache is an accessor for the whole cache, hardwired to the cache. I would like the processes, the packages, etc, to be agnostic of the physical layout/location. A package will be in user space in ``workspaces``, or in the local flow. The cache won't exist there for that package.

**Comment by jgsogo** [comment]
I understand, but then `ClientCache` should be decomposed in a `PackagesFolder` (coupled to the cache directory) and the rest of the data.

And then use the `PackagesFolder` class everywhere instantiated with the proper directory.

**Comment by jgsogo** [comment]
I find myself thinking about this feature from time to time and I really want to give it a try. So let me rephrase it (once more) in order to understand it better.

The objective could be to implement some kind of delegation logic that allows mixing different locations for packages at the same time (on the same ``conan <command>`` run):
- when using workspaces we would want to use some packages from a local directory and other ones from the cache.
- use a temporary folder to do a one-off action over a package (like inspecting the same package from different remotes)
- perform a ``conan download`` to a specific folder
- ...

Some concerns:
- I'm not sure if these _local caches_ should have or not the same directory structure as the current one:
   - For temporary inspection, download,... where several versions and package_ids could coexists I'm sure the current layout will work.
   - But for workspaces where the developer will be working with a clone of the repo, the layout should resemble the one of the local development workflow.

What I will keep from the current behavior:
- Upload to remotes should be possible only from the _official_ location of the cache (so you need to _export_ to that cache, and the _upload_).

----
So the implementation that is taking shape in my head is: 
- From the code perspective, all should be transparent, just one object of the ``ClientCache``, all the logic should be contained inside that object.
- Implement **different layouts** for packages given a location: current cache, layout for workspaces, temporary layout,... with different lifecycles.
- ``ClientCache`` will resolve to filesystem every conan reference prioritizing those layouts depending on the command, some inputs, configuration,... (some of them may not exists).
- When a package is retrieved from remotes, it should be assigned to one of those layouts and managed by it.

...keep thinking about it, but feedback is more than welcome. Will it help only to the _Workspaces feature_ or something else could take advantage of it?
